contracts/libraries/VaultLifecycleSTETH.sol

```diff
@@ -499,6 +499,62 @@ library VaultLifecycleSTETH {
         }
     }
 
+    /**
+     * @notice Gets stETH for direct stETH withdrawals, converts wstETH/ETH to stETH if not enough stETH
+     * @param steth is the address of steth
+     * @param wstEth is the address of wsteth
+     * @param amount is the amount to withdraw
+     * @return amount of stETH to transfer to the user, this is to account for rounding errors when unwrapping wstETH
+     */
+    function withdrawStEth(
+        address steth,
+        address wstEth,
+        uint256 amount
+    ) external returns (uint256) {
+        // 3 different scenarios for withdrawing stETH directly
+        // Scenario 1. We hold enough stETH to satisfy withdrawal. Send it out directly
+        // Scenario 2. We hold enough stETH + wstETH to satisy withdrawal. Unwrap wstETH then send it
+        // Scenario 3. We hold enough stETH + wstETH + ETH satisfy withdrawal. Unwrap wstETH, wrap ETH then send it
+        uint256 stethBalance = IERC20(steth).balanceOf(address(this));
+        if (stethBalance >= amount) {
+            // Can send out the stETH directly
+            return amount; // We return here if we have enough stETH to satisfy the withdrawal
+        } else {
+            // If amount > stethBalance, send out the entire stethBalance and check wstETH and ETH
+            amount = amount.sub(stethBalance);
+        }
+        uint256 wstethBalance = IWSTETH(wstEth).balanceOf(address(this));
+        stethBalance = IWSTETH(wstEth).getStETHByWstETH(wstethBalance);
+        if (stethBalance >= amount) {
+            wstethBalance = IWSTETH(wstEth).getWstETHByStETH(amount);
```

Comment:
My understanding from reading the steth contracts are that it's these functions that are adding to the rounding errors. Each conversion rounds up the amounts (muldiv rounded up), so when it's `.add(3)` in the tests, it probably has gone through:

1. getStETHByWstETH
2. getWstETHByStETH
3. unwrap

One way to save gas and lessen this rounding error is to do the muldiv ourselves, eg:

```solidity
uint stEthPerToken = IWSTETH(wstEth).stEthPerToken();
stethBalance = (wstethBalance * stEthPerToken)/10**18; // this rounds down by default
wstethBalance = (amount * 10**18) / stEthPerToken;

IWSTETH(wstEth).unwrap(wstethBalance);
```

Reply:
Good idea, `stEthPerToken` saves gas
In the background it makes an additional call to the stETH, so its cheaper to get the `getTotalShares` and `getTotalPooledEther` directly from stETH and do the calculations

```
uint256 totalShares = ISTETH(steth).getTotalShares();
uint256 totalPooledEther = ISTETH(steth).getTotalPooledEther();
stethBalance = wstethBalance.mul(totalPooledEther).div(totalShares);

wstethBalance = amount.mul(totalShares).div(totalPooledEther);
```